File: week9/박현국/괄호_회전하기.java
package 박현국;

import java.util.*;

class 괄호_회전하기 {
    public int solution(String s) {
        int answer = 0;

        for (int i = 0; i < s.length(); i++) {
            if (isValid(s)) {
                answer++;
            }
            s = rotate(s);
        }

        return answer;
    }

    private String rotate(String s) {
        return s.substring(1, s.length()) + s.substring(0, 1);
    }

    private boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '[' || s.charAt(i) == '{' || s.charAt(i) == '(') {
                stack.push(s.charAt(i));
            } else if (s.charAt(i) == ']') {
                if (stack.isEmpty()) {
                    return false;
                }

                if (stack.pop() != '[') {
                    return false;
                }
            } else if (s.charAt(i) == '}') {
                if (stack.isEmpty()) {
                    return false;
                }

                if (stack.pop() != '{') {
                    return false;
                }
            } else if (s.charAt(i) == ')') {
                if (stack.isEmpty()) {
                    return false;
                }

                if (stack.pop() != '(') {
                    return false;
                }
            }
        }

        if (! stack.isEmpty()) {
            return false;
        }
        return true;
    }
}

File: week9/한지희/괄호_회전하기.kt
import java.util.*

class RotateBracket {
    fun solution(s: String): Int {
        var answer = 0

        for (i in s.indices) {
            if (isBracketString(s.substring(i, s.length) + s.substring(0, i))) answer++
        }

        return answer
    }

    private fun isBracketString(s: String): Boolean {
        val openBracket = listOf('(', '[', '{')
        val closeBracket = listOf(')', ']', '}')
        val bracketStack = Stack<Char>()

        s.forEach { char ->
            if (char in openBracket) bracketStack.push(char)
            else {
                if (bracketStack.isEmpty()) return false
                if (char != closeBracket[openBracket.indexOf(bracketStack.peek())]) return false
                bracketStack.pop()
            }
        }
        return bracketStack.isEmpty()
    }

    fun bestSolution(s: String): Int {
        return s.indices.count { isBracketString(s.substring(it, s.length) + s.substring(0, it)) }
    }
}